route-rs-runtime/src/link/mod.rs

```diff
@@ -28,9 +27,9 @@ pub mod primitive;
 pub mod utils;
 
 /// All Links communicate through streams of packets. This allows them to be composable.
-pub type PacketStream<Input> = Box<dyn Stream<Item = Input, Error = ()> + Send>;
+pub type PacketStream<Input> = Box<dyn futures::Stream<Item = Input> + Send + Unpin>;
```

Comment:
Can you describe why `PacketStream`s are `Unpin`? Is this just explicitly stating that packets may be moved/copied across threads/cores? I thought [most types implement `Unpin` by default](https://doc.rust-lang.org/std/marker/trait.Unpin.html), but a generic type needs to be marked?

Reply:
Maybe without `Unpin` we would need to enforce that every packet is `Pin`ned, and that's not a restriction we want to make? https://rust-lang.github.io/async-book/04_pinning/01_chapter.html#how-to-use-pinning

Reply:
This isn't that the packets need to be Unpin, it is the object that provides the Stream must be unpin. Unpin means we can freely pin and then unpin the object. Structs wont implement unpin by default because they may contain pointers and such that could move. They could provide a custom implementation if that were the case, but since the ingressor and egressor own all their components, it isn't an issue to just add the trait to get the included 'provided' methods. It is because when we call poll_next on that stream object, we need to pin it in place so the futures work correctly.  AFAIK, this is because the runtime needs to know that the object doesn't move out from underneath it when it is trying to reference the future. The poll_next function can't move on the runtime otherwise its calls would become invalidated it the function pointer they stored no longer pointed at the poll function cause the object moved in memory.  Something like that, I think the withoutboats talk may be more illuminating.  But in essence, we need our objects that implement packetstream to impl unpin so that I can shove them into a Pin object to call their poll_next() function.  Checkout the function sig of poll_next to see why.

Reply:
I think unpin is kind of a bad name, but I'm not sure if I can think of anything better.

Reply:
Okay I think this will be informative: Take a look at ClassifyIngessor's Future impl
`impl<'a, C: Classifier> Future for ClassifyIngressor<'a, C> {
    type Output = ();

    /// Same logic as QueueEgressor, except if any of the channels are full we
    /// await that channel to clear before processing a new packet. This is somewhat
    /// inefficient, but seems acceptable for now since we want to yield compute to
    /// that egressor, as there is a backup in its queue.
    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
        let ingressor = Pin::into_inner(self);
        loop {
            for (port, to_egressor) in ingressor.to_egressors.iter().enumerate() {
                if to_egressor.is_full() {
                    park_and_notify(&ingressor.task_parks[port], cx.waker().clone());
                    return Poll::Pending;
                }
            }

            //TODO: Standardize in_stream, input_stream, and stream to one name
            let packet_option: Option<C::Packet> =
                ready!(Pin::new(&mut ingressor.input_stream).poll_next(cx));

            match packet_option {
                None => {
                    for to_egressor in ingressor.to_egressors.iter() {
                        to_egressor
                            .try_send(None)
                            .expect("ClassifyIngressor::Drop: try_send to_egressor shouldn't fail");
                    }
                    for task_park in ingressor.task_parks.iter() {
                        die_and_notify(&task_park);
                    }
                    return Poll::Ready(());
                }
                Some(packet) => {
                    let class = ingressor.classifier.classify(&packet);
                    let port = (ingressor.dispatcher)(class);
                    if port >= ingressor.to_egressors.len() {
                        panic!("Tried to access invalid port: {}", port);
                    }
                    if let Err(err) = ingressor.to_egressors[port].try_send(Some(packet)) {
                        panic!(
                            "Error in to_egressors[{}] sender, have nowhere to put packet: {:?}",
                            port, err
                        );
                    }
                    unpark_and_notify(&ingressor.task_parks[port]);
                }
            }
        }
    }
}`

The first interesting line: `        let ingressor = Pin::into_inner(self);`
This ingressor, when the runtime calls it via Poll(..), has a Pinned version of the ingressor for self.  So to actually get at the internal values, we need to unpin it, so then we can refer to our methods and member variables as usual. If this were not an unpin struct, ie it was self referential, it wouldn't work. If it is self-referential or something like that, once you Pin it in place, you say "This can't be moved without invalidating something", which in the this example would be the internal pointer to itself.  Prior to Pin I don't think being self referential was something the rust lang supported. 

The second line: `ready!(Pin::new(&mut ingressor.input_stream).poll_next(cx));` Is the same thing but in reverse. Before we can call poll_next on our Stream object, we need to tell the computer to pin the Stream in place, so to satisfy the requirements of poll_next and poll, where self is `self: Pin<&mut Self>`

Reply:
> Structs wont implement unpin by default because they may contain pointers and such that could move. They could provide a custom implementation if that were the case, but since the ingressor and egressor own all their components, it isn't an issue to just add the trait to get the included 'provided' methods.

It feels like the compiler should be able to do some kind of analysis on structs to determine whether or not this is the case, but i don't wanna yell at the core team. I'm just happy `async` is stable.

That example was pretty informative. I'm going to watch withoutboats' talk on `Pin` to refresh my knowledge, but I'm going to trust your judgement 👌